### ISSUE TITLE
Text width and height

### ISSUE THREAD

**Issue by BernhardWerner** [CindyJS/CindyJS#764]
When writing text I would like to draw a rectangle or any other shape around it. (For buttons, nodes in a graph, etc.) To do so properly, I would like to have a function that outputs the width (and another for the height) of text. I.e., something like `textwidth(string, size)` would return the width (in CindyJS units) of `string` if drawn using `drawtext` using `size`.

**Comment by kortenkamp** [comment]
Please, if anyone is considering implementing this - such a function would imply that we only have one port. In classic Cinderella the scripts are executed only once for all ports.

I am open to discussing alternative approaches (for example port-specific draw events), but this needs a good design.

**Comment by montaga** [comment]
It would be thinkable that a `measuretext( text, +modifiers of drawtext...)` could return the size in pixels (internally such a function already exists that should even work well together with KaTeX-rendered formulas). If divided by the port-dependent function `screenresolution()` its first component would correspond the width in CindyJS units.

**Comment by kortenkamp** [comment]
‘screenresolution()‘ is not port-dependent...

**Comment by montaga** [comment]
Oh, I was not aware of that. According to the reference "If several Euclidean views for the same construction are open, the function [`screenresolution()`] will return the maximum of all individual screen resolutions.".

**Comment by BernhardWerner** [comment]
This was solved/Fulfilled with the new `pixelsize` function. But if I see it correctly, it doesn't work properly for KaTeX formulas:
E. g., the code
```
testText = "$\sum 1234$";
testSize = 20;

a = pixelsize(testText, size -> testSize)_1 / screenresolution();

box = drawtext([0,0], testText, size -> testSize);
b = dist(box_1, box_2);

draw([0,-1], [a,-1], size -> 3, color -> (1,0,0));
draw([0,-2], [b,-2], size -> 3, color -> (0,0,1));
```
as the draw script shows that `a` is much bigger than `b`. From what I can see, `pixelsize` gives the width of the string when it doesn't get converted to TeX. Any idea why, @kortenkamp ?

IIRC, it did work when it was first added to the main branch. And I only noticed the wrong behaviour after I added my text outline for which I did have to change the KaTeX plugin and KaTeX itself a little. But I cannot see how any of my changes would affect the measurement or whether Cindy recognises a TeX string.

**Comment by kortenkamp** [comment]
Hm. I can look into it. But yes, this was working before…

**Comment by BernhardWerner** [comment]
Did a quick check. Taking the current main branch and moving the HEAD back to
- commit `a8e37e8a` where pixelsize was introduced,
- commit `cfb132a6` where Michael added the bugfix to drawtext or
- commit `f2a8ec05` right before I added any of the text outline code

all lead to version where pixelsize doesn't work. Which is good for me, but doesn't help. Will investigate more...

**Comment by BernhardWerner** [comment]
I maybe have a fix:

Firstly, I'm still not sure how `pixelsize` worked in the first place. If I see it correctly, it loads KaTeX if necessary, but then only uses the function `measureNoRendererCanvas` defined in the same file `OpDrawings.js`. Even if KaTeX is loaded, it is never used, I think.

Secondly, I added a dedicated measure function to the KaTeX plugin and then extended the CindyJS API such that this measure function is swapped in if necessary. I.e. it works just like `textRendererCanvas` and `textRendererHtml`. And it mostly works! There's still an issue when KaTeX isn't fully loaded, so it doesn't work in the init script on TeX strings. (Cf. #932.) Would this be an acceptable approach as soon as it works?

Thirdly, I cannot make a commit because of the error message
```bash
✖ eslint --cache --fix:
4777:21  error  'setMeasureNoRendererCanvas' is not defined  no-undef
husky - pre-commit hook exited with code 1 (error)
```
I defined and set up this function the same way as `setTextRendererHtml` in `OpDrawings.js`. Concretely:
```JavaScript
let measureNoRendererCanvas = defaultMeasureNoRendererCanvas;

function setMeasureNoRendererCanvas(fct) {
    measureNoRendererCanvas = fct;
}
```
right after the definition of `defaultMeasureNoRendererCanvas`. Which is the old `measureNoRendererCanvas` function.
Is there a way to find out what exactly the issue is? Or to tell eslint/husky or whatever is doing the checks to... recognize this new function?

**Comment by BernhardWerner** [comment]
Alright, I managed to remember the git commit option `--no-verify` so you can at least look at this try of a solution in the branch `pixelsize-fix`. The questions from above still stand, though:
1. How did the original implementation of `pixelsize` ever work without using KaTeX to parse the string?
2. Would the version in the branch `pixelsize-fix` be fine per se? If yes, how could we ever get `pixelsize` to work in the init script if it needs KaTeX?
3. What's eslint's problem?

**Comment by kortenkamp** [comment]
@BernhardWerner – thanks for submitting. I have no objections at all. To answer your question:

1. I am not sure that it ever worked.
2. I am happy with anything that works better than before.
3. Maybe @strobelm can answer that… I never tried to understand the overcomplex workflow here. I also used `--no-verify` sometimes just to be able to continue to work on solving problems…

The big open question is how we can integrate a new KaTeX version. Or replace it by something else.

**Comment by BernhardWerner** [comment]
@kortenkamp, then I'll take a final look tomorrow and do a pull request.

About KaTeX: We could somewhat easily integrate it in CindyJS itself -- instead of having it as a plugin -- so that it properly loads everything before the init script runs. That would solve the problem here of pixelsize not working in the init script. But I'm not sure whether that's a good and/or desired solution.

**Comment by kortenkamp** [comment]
No, that is not something we want to do. The problem I mean is that we use an ancient version of KaTeX with custom modifications that we cannot backport to the current release of KaTeX. I will look into that someday.

**Comment by kortenkamp** [comment]
Just to answer your question about 'setMeasureNoRendererCanvas' – you forgot to import it in 'Operator.js'. I fixed this for the release of v0.8.32
